src/main/java/com/google/api/pathtemplate/PathTemplate.java

```diff
@@ -1015,6 +1013,20 @@ public String encode(String... values) {
     return builder.build();
   }
 
+  private static boolean isSegmentBeginOrEndInvalid(String seg) {
+    if (seg.length() == 1 && COMPLEX_DELIMITER_PATTERN.matcher(seg).find()) {
+      return true;
+    }
+    //A segment like .{well}-{known} or {well}-{known}. is invalid.
+    //A segment like .well-known or .well-{known} will be considered a literal hence is valid
+    return (COMPLEX_DELIMITER_PATTERN.matcher(seg.substring(0, 1)).find()
```

Comment:
Minor: please consider reversing the order of conditions in `&&` to make sure that lighter ones are executed first (and potentially not needing to go to the much heavier regexp matching portion if the simpler ones are already false)

Reply:
The order cannot be changed since a segment may only contain one non-delimiter character, so we have to test the regex first.

Reply:
I'm a big confused. In the original code, each `&&` segment also had `seg.length() > 1` which was guaranteeing that seg is at lest 2 characters long, so both `seg.charAt(1) == '{'` and `seg.charAt(seg.length() - 2) == '}'` would not fail. Instead of removing `seg.length() > 1` from both segments, we could just put it infront of the whole return block condition, and that wuold allow testing for `seg.charAt(1) == '{'` and `seg.charAt(seg.length() - 2) == '}'` prior the regexp parts without risk getting out of bounds exception which is preferable, since  these are more lightweight operations and if they are false we would not even get into regexp part. 

But all that is micro-optimization stuff, so I'm fine with how it is now.
On a side note, the current code would fail if seg is an empty string, but I'm not sure if it is even possible.

Reply:
Yes it would fail on empty String but that's not possible because of the trimming [here](https://github.com/googleapis/api-common-java/blob/6a8a0a44fbf942ea6224ab3d8d4b3bcc6c79c5b4/src/main/java/com/google/api/pathtemplate/PathTemplate.java#L881)